### PR TITLE
Set of build fixes for ## 3580, 3574.

### LINKED ISSUE #3574: Polymorphic type reference error during build
### System Information

This error occurs after make j*:

```

Scanning dependencies of target opencv_perf_imgproc
[ 37%] Building CXX object modules/imgproc/CMakeFiles/opencv_perf_imgproc.dir/perf/opencl/perf_3vs4.cpp.o
[ 37%] Building CXX object modules/cudaarithm/CMakeFiles/opencv_test_cudaarithm.dir/test/test_event.cpp.o
[ 37%] Building CXX object modules/cudaarithm/CMakeFiles/opencv_perf_cudaarithm.dir/perf/perf_element_operations.cpp.o
/home/ubuntu/opencv_contrib/modules/cudaarithm/test/test_event.cpp: In member function ‘void opencv_test::{anonymous}::AsyncEvent_Timing_Test::UnsafeTestBody()’:
/home/ubuntu/opencv_contrib/modules/cudaarithm/test/test_event.cpp:94:30: warning: catching polymorphic type ‘class cv::Exception’ by value [-Wcatch-value=]
   94 |         catch (cv::Exception ex) {


```

The warning message you're seeing is related to catching a polymorphic type by value instead of by reference. In C++, it's a good practice to catch exceptions by reference rather than by value to preserve the exception type information and to avoid object slicing.

Here's how you could fix this warning:

Change this line:
`catch (cv::Exception ex) {`

To:
`catch (const cv::Exception& ex) {`

### Detailed description

By making this change, you're catching the exception by reference, which should eliminate the warning message. Now go back to your source code, find the file test_event.cpp, locate line 94, and make the suggested change. Save the file and re-compile your project. This should resolve the warning you were seeing.

### Steps to reproduce

make j*

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

### LINKED ISSUE #3580: compile failure due to <Type> syntax
##### System information (version)

- OpenCV => 4.x
- Operating System / Platform => ubuntu 22.04
- Compiler => gcc 11.4

##### Detailed description
On trying to compile 4.x with CUDA, this issue comes up: 
opencv_contrib/modules/xphoto/src/oilpainting.cpp:61:34: error: expected ‘)’ before ‘&’ token
   61 |     ParallelOilPainting<Type>(Mat& img, Mat &d, Mat &iLuminance, int r,int k) :

##### Steps to reproduce

My CMAKE flags were: 
      -D CUDNN_LIBRARY=usr/lib/x86_64-linux-gnu/libcudnn.so \
      -D CUDNN_INCLUDE_DIR=/usr/include \
      -D CMAKE_CXX_STANDARD=17 \
      -D OPENCV_GENERATE_PKGCONFIG=YES \
      -D CMAKE_CXX_STANDARD_REQUIRED=ON \
      -D CMAKE_BUILD_TYPE=RELEASE \
      -D CMAKE_INSTALL_PREFIX="$install_path" \
      -D OPENCV_EXTRA_MODULES_PATH="$contrib_path/modules" \
      -D WITH_CUDA=ON \
      -D WITH_CUDNN=ON \
      -D WITH_GTK=ON \
      -D WITH_VTK=OFF \
      -D OPENCV_DNN_CUDA=ON \
      -D ENABLE_FAST_MATH=1 \
      -D CUDA_FAST_MATH=1 \
      -D CUDA_ARCH_BIN="$cuda_arch_bin" \
      -D WITH_CUBLAS=1 \
      -D WITH_LIBV4L=ON \
      -D WITH_OPENGL=ON \
      -D WITH_TBB=ON \
      -D WITH_QT=ON \
      -D OPENCV_ENABLE_NONFREE=ON \
      -D INSTALL_PYTHON_EXAMPLES=OFF \
      -D BUILD_opencv_java=OFF \
      -D BUILD_opencv_python=OFF \
      -D INSTALL_C_EXAMPLES=OFF \
      -D INSTALL_PYTHON_EXAMPLES=OFF \
      -D BUILD_EXAMPLES=OFF \
      -D CMAKE_CXX_FLAGS=${SIMD_FLAG}

The <Type> seems to break cpp compiles for C++ 17, and 23. Removing "Type " addresses the issue.

##### Issue submission checklist

 - [ X] I report the issue, it's not a question
   
 - [X ] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found any solution
   
 - [ X] I updated to the latest OpenCV version and the issue is still there
   
 - [ X] There is reproducer code and related data files: videos, images, onnx, etc
   

The same issue is true for opencv_contrib/modules/datasets/src/tinyxml2/tinyxml2.h: At global scope:
opencv_contrib/modules/datasets/src/tinyxml2/tinyxml2.h:215:25: error: expected unqualified-id before ‘)’ token
  215 |     DynArray< T, INIT >() {

### PR DESCRIPTION
**Pull request by asmorkalov**
Fixes #3580
Fixes #3574.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by mshabunin** [comment]
Windows build on buildbot has failed:
```
C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp(1216): error C2131: expression did not evaluate to a constant [C:\build\precommit-contrib_windows64\build\modules\face\opencv_face.vcxproj]
  C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp(1216): note: failure was caused by non-constant arguments or reference to a non-constant symbol
  C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp(1216): note: see usage of 'l'
C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp(1217): error C3863: array type 'double [l]' is not assignable [C:\build\precommit-contrib_windows64\build\modules\face\opencv_face.vcxproj]
C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp(1218): error C3863: array type 'double [l]' is not assignable [C:\build\precommit-contrib_windows64\build\modules\face\opencv_face.vcxproj]
C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp(1314): error C2665: 'cv::max': none of the 6 overloads could convert all the argument types [C:\build\precommit-contrib_windows64\build\modules\face\opencv_face.vcxproj]
  C:\build\precommit-contrib_windows64\4.x\opencv\modules\core\include\opencv2/core/mat.hpp(3766): note: could be 'cv::MatExpr cv::max(double,const cv::Mat &)' (compiling source file C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp)
  C:\build\precommit-contrib_windows64\4.x\opencv\modules\core\include\opencv2/core/mat.hpp(3765): note: or       'cv::MatExpr cv::max(const cv::Mat &,double)' (compiling source file C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp)
  C:\build\precommit-contrib_windows64\4.x\opencv\modules\core\include\opencv2/core/mat.hpp(3764): note: or       'cv::MatExpr cv::max(const cv::Mat &,const cv::Mat &)' (compiling source file C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp)
  C:\build\precommit-contrib_windows64\4.x\opencv_contrib\modules\face\src\facemarkLBF.cpp(1314): note: while trying to match the argument list '(double, double [l])'
```

File: modules/face/src/facemarkLBF.cpp [hunk 1]
```diff
@@ -1212,8 +1212,9 @@ Mat FacemarkLBFImpl::Regressor::supportVectorRegression(
     double Gnorm1_init = -1.0; // Gnorm1_init is initialized at the first iteration
     std::vector<double> beta(l);
     std::vector<double> QD(l);
+    std::vector<double> lambda(l);
+    std::vector<double> upper_bound(l);
```
Comment (mshabunin) [inline modules/face/src/facemarkLBF.cpp:1216]: Mentioned commit in scikit-learn also changes `#define GETI(i) (i)` and adds initialization of `lambda` and `upper_bound` arrays: https://github.com/scikit-learn/scikit-learn/blob/dcebbc4c5c97f886a57b0684a5e943f94db1bd58/sklearn/svm/src/liblinear/linear.cpp#L1073-L1089
And assuming we have L2R_L1LOSS_SVR_DUAL case here, single-element arrays make sense, but `GETI` macro is wrong.

Alternatively there is another source of original code where `GETI(i)` is `0` and arrays of size 1 works well: https://github.com/cjlin1/liblinear/blob/8dc206b782e07676dc0d00678bedd295ce85acf3/linear.cpp#L1127

**Comment by asmorkalov** [comment]
@mshabunin I reverted changes in `facemarkLBF.cpp`. Let's fix it with the next PR.

**Review by mshabunin — APPROVED** [review]
